internal/v5/auth.go

```diff
@@ -64,10 +76,120 @@ func (h *ReqHandler) authorize(req *http.Request, acl []string, alwaysAuth bool,
 	}
 
 	// Macaroon verification failed: mint a new macaroon.
-	m, err := h.newMacaroon()
+	// We need to deny access for opAccessCharmWithTerms operations because they
+	// may require more specific checks that terms and conditions have been
+	// satisfied.
+	m, err := h.newMacaroon(checkers.DenyCaveat(opAccessCharmWithTerms))
+	if err != nil {
+		return authorization{}, errgo.Notef(err, "cannot mint macaroon")
+	}
+
+	// Request that this macaroon be supplied for all requests
+	// to the whole handler.
+	// TODO use a relative URL here: router.RelativeURLPath(req.RequestURI, "/")
+	cookiePath := "/"
+	return authorization{}, httpbakery.NewDischargeRequiredErrorForRequest(m, cookiePath, verr, req)
+}
+
+// authorizeEntityAndTerms is similar to the authorize method, but
+// in addition it also checks if the entity meta data specifies
+// and terms and conditions that the user needs to agree to. If so,
+// it will require the user to agree to those terms and conditions
+// by adding a third party caveat addressed to the terms service
+// requiring the user to have agreements to specified terms.
+func (h *ReqHandler) authorizeEntityAndTerms(req *http.Request, entityIds []*router.ResolvedURL) (authorization, error) {
+	logger.Infof(
+		"authorize entity and terms, auth location %q, terms location %q, path: %q, method: %q, entities: %#v",
+		h.Handler.config.IdentityLocation,
+		h.Handler.config.TermsLocation,
+		req.URL.Path,
+		req.Method,
+		entityIds)
+
+	if len(entityIds) == 0 {
+		return authorization{}, errgo.WithCausef(nil, params.ErrUnauthorized, "entity id not specified")
+	}
+
+	ACLs := make([][]string, len(entityIds))
+	requiredTerms := make(map[string]struct{})
+	allowAuth := true
+	for i, entityId := range entityIds {
```

Comment:
This is quite a long function and this is perhaps a coherent section of it that might be removed into a separate function, making its purpose and side-effects more clear.

Something like:

    // entityAuthInfo returns authorization on the entities with the given ids.
    // If public is true, no authorization is required, otherwise acls holds
    // an entry for each id with the corresponding ACL for each entity,
    // and requiredTerms holds entries for all required terms.
    func (h *ReqHandler) entityAuthInfo(entityIds []*router.ResolvedURL) (public bool, acls [][]string, requiredTerms map[string]bool, err error)

perhaps?